prune.go

```diff
@@ -129,122 +135,194 @@ func txnsPruneC(txnsName string) string {
 // with a bit of luck something like this will one day be part of
 // mgo/txn.
 func PruneTxns(db *mgo.Database, txns *mgo.Collection) error {
-	present := struct{}{}
-
-	// Load the ids of all completed txns and all collections
-	// referred to by those txns.
-	//
-	// This set could potentially contain many entries, however even
-	// 500,000 entries requires only ~44MB of memory. Given that the
-	// memory hit is short-lived this is probably acceptable.
-	txnIds := make(map[bson.ObjectId]struct{})
-	collNames := make(map[string]struct{})
+	workingSetName := txns.Name + ".prunetemp"
+	workingSet := db.C(workingSetName)
+	defer workingSet.DropCollection()
 
-	var txnDoc struct {
-		Id  bson.ObjectId `bson:"_id"`
-		Ops []txn.Op      `bson:"o"`
+	// Load the ids of all completed and aborted txns into a separate
+	// temporary collection.
+	logger.Debugf("loading all completed transactions")
+	matchCompleted := bson.M{"$match": bson.M{"s": bson.M{"$in": []int{taborted, tapplied}}}}
+	pipe := txns.Pipe([]bson.M{
+		matchCompleted,
+		{"$project": bson.M{"_id": 1}},
+		{"$out": workingSetName},
+	})
+	var dummy bson.D
+	if err := runPipe(pipe, &dummy); err != nil {
+		return fmt.Errorf("reading completed txns: %v", err)
 	}
 
-	logger.Debugf("loading all completed transactions")
-	completed := bson.M{
-		"s": bson.M{"$in": []int{taborted, tapplied}},
-	}
-	iter := txns.Find(completed).Select(bson.M{"_id": 1, "o": 1}).Iter()
-	for iter.Next(&txnDoc) {
-		txnIds[txnDoc.Id] = present
-		for _, op := range txnDoc.Ops {
-			collNames[op.C] = present
-		}
+	// Load all collections which have completed transactions.
+	logger.Debugf("loading collections with transactions")
+	pipe = txns.Pipe([]bson.M{
+		matchCompleted,
+		{"$project": bson.M{"_id": 0, "c": "$o.c"}},
```

Comment:
I take it this is "give me only things in the o.c field as just the value c"
and then "turn the arrays of C into a single top level list of C" objects
and then "aggregate all of the separate C objects so that we have only unique c values".

Given that you don't have interesting documents (just a single field), is there something cheaper than "$group"? Maybe not.

Given all this work only gives you the collection names. I'm curious why do it this way instead of just iterating all known collections and looking for 'txn-queue' fields.